src/group_theory/submonoid/center.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import group_theory.submonoid.operations
+
+/-!
+# Centers of magmas and monoids
+
+## Main definitions
+
+* `set.center`: the center of a magma
+* `submonoid.center`: the center of a monoid
+* `set.add_center`: the center of an additive magma
+* `add_submonoid.center`: the center of an additive monoid
+
+We show `subgrup.center` and `subsemiring.center` and `subring.center` in other files.
+-/
+
+variables {M : Type*}
+
+namespace set
+
+variables (M)
+
+/-- The center of a magma. -/
+@[to_additive add_center /-" The center of an additive magma. "-/]
+def center [has_mul M] : set M := {z | ∀ m, m * z = z * m}
+
+@[to_additive mem_add_center]
+lemma mem_center_iff [has_mul M] {z : M} : z ∈ center M ↔ ∀ g, g * z = z * g := iff.rfl
+
+@[simp, to_additive zero_mem_add_center]
+lemma one_mem_center [mul_one_class M] : (1 : M) ∈ set.center M := by simp [mem_center_iff]
+
+@[simp]
+lemma zero_mem_center [mul_zero_class M] : (0 : M) ∈ set.center M := by simp [mem_center_iff]
+
+variables {M}
+
+@[simp, to_additive add_mem_add_center]
+lemma mul_mem_center [semigroup M] {a b : M}
+  (ha : a ∈ set.center M) (hb : b ∈ set.center M) : a * b ∈ set.center M :=
+λ g, by rw [mul_assoc, ←hb g, ← mul_assoc, ha g, mul_assoc]
+
+@[simp, to_additive neg_mem_add_center]
+lemma inv_mem_center [group M] {a : M} (ha : a ∈ set.center M) : a⁻¹ ∈ set.center M :=
+λ g, by rw [← inv_inj, mul_inv_rev, inv_inv, ← ha, mul_inv_rev, inv_inv]
+
+@[simp]
+lemma add_mem_center [distrib M] {a b : M}
+  (ha : a ∈ set.center M) (hb : b ∈ set.center M) : a + b ∈ set.center M :=
+λ c, by rw [add_mul, mul_add, ha c, hb c]
+
+@[simp]
+lemma neg_mem_center [ring M] {a : M} (ha : a ∈ set.center M) : -a ∈ set.center M :=
+λ c, by rw [←neg_mul_comm, ha (-c), neg_mul_comm]
+
+@[to_additive subset_add_center_add_units]
+lemma subset_center_units [monoid M] :
+  (coe : units M → M) ⁻¹' center M ⊆ set.center (units M) :=
+λ a ha b, units.ext $ ha _
+
+lemma center_units_subset [group_with_zero M] :
+  set.center (units M) ⊆ (coe : units M → M) ⁻¹' center M :=
+λ a ha b, begin
+  obtain rfl | hb := eq_or_ne b 0,
+  { rw [zero_mul, mul_zero], },
+  { exact units.ext_iff.mp (ha (units.mk0 _ hb)) }
+end
+
+/-- In a group with zero, the center of the units is the preimage of the center. -/
+lemma center_units_eq [group_with_zero M] :
+  set.center (units M) = (coe : units M → M) ⁻¹' center M :=
+subset.antisymm center_units_subset subset_center_units
+
+@[simp]
+lemma inv_mem_center' [group_with_zero M] {a : M} (ha : a ∈ set.center M) : a⁻¹ ∈ set.center M :=
+begin
+  obtain rfl | ha0 := eq_or_ne a 0,
+  { rw inv_zero, exact zero_mem_center M },
+  lift a to units M using ha0,
+  rw ←units.coe_inv',
+  exact center_units_subset (inv_mem_center (subset_center_units ha)),
+end
```

Comment:
I'm sure there's a nicer way to prove `inv_mem_center`, but I figured the lemmas about the units were worth having anyway.